src/com/google/enterprise/adaptor/database/DatabaseAdaptor.java

```diff
@@ -266,14 +267,12 @@ public void getDocIds(DocIdPusher pusher) throws IOException,
   void addMetadataToRecordBuilder(DocIdPusher.Record.Builder builder,
       ResultSet rs) throws SQLException {
     ResultSetMetaData rsMetaData = rs.getMetaData();
-    int numberOfColumns = rsMetaData.getColumnCount();
-    for (int i = 1; i < (numberOfColumns + 1); i++) {
-      String columnName = rsMetaData.getColumnLabel(i);
-      Object value = rs.getObject(i);
-      String key = metadataColumns.getMetadataName(columnName);
-      if (key != null) {
-        builder.addMetadata(key, "" + value);
-      }
+    MetadataColumns metadataCols = (metadataColumns == null)
+        ? new MetadataColumns(rsMetaData) : metadataColumns;
+    for (String columnName : metadataCols.keySet()) {
```

Comment:
As discussed, if we use entrySet instead of keySet, we can make MetadataColumns extend AbstractMap at our leisure, remove some of the helper methods, and restore magic in a get method if desired later.